### DISCUSSION TITLE
Tree shaking and barrel file from libraries

### DISCUSSION THREAD

**Discussion by MLiapkalo** [nrwl/nx#28714] [Questions]
To start with, here's a minimal repo where you can see the behaviour I'm going to describe: [repo](https://github.com/MLiapkalo/mf-di)
And here's the nx graph of workspace from repo:

<img width="987" alt="image" src="https://github.com/user-attachments/assets/da8a8474-2b7f-4095-b0bc-7ffa0006c77d">

My question is how can I perform tree shaking on barrel files from data-access libraries? Running webpack-bundle-analyzer(run `nx analyze host` inside the workspace from provided repo) shows that everything from data-access ends up in initial bundle(or am I misinterpreting the results?). I understand that barrel files from data-access libraries re-export every module the lib has, but I was expecting tree shaking come into action.

Output of `nx analyze host`:

<img width="1705" alt="image" src="https://github.com/user-attachments/assets/3bf3b16e-2834-40b1-9c09-412282b340ec">

Both `books-data-access` and `users-data-access` libraries have `sideEffects: false` in their `package.json` files and I use only a single module from each data-access lib inside host.

**Comment by pumano** [comment]
bro for proper tree-shaking you need to use named export `export { something } from 'something';` , no namespace export via `export * from 'something';` and you should see proper tree-shaking.

**Reply by MLiapkalo** [reply]
Hi, @pumano . I've tested your suggestion. Using named exports yields the same result as namespace export does.

**Reply by pumano** [reply]
It's only exported, but not imported somwhere?

**Reply by MLiapkalo** [reply]
only a couple of re-exported modules is used, others have no usages but still make it to the intial bundle
```
export { UsersApiService } from './lib/users-api.service'; // USED
export { UsersApiConfiguration } from './lib/users-api-configuration'; // USED
export { User } from './lib/models/user'; // unused
export { apiV1UsersIdPut } from './lib/fn/users/api-v-1-users-id-put' // unused
export { apiV1UsersPost } from './lib/fn/users/api-v-1-users-post' // unused
export { apiV1UsersIdGet } from './lib/fn/users/api-v-1-users-id-get' // unused
export { apiV1UsersIdDelete } from './lib/fn/users/api-v-1-users-id-delete' // unused
export { apiV1UsersGet$JsonV10 } from './lib/fn/users/api-v-1-users-get-json-v-10' // USED
export { apiV1UsersGet$PlainV10 } from './lib/fn/users/api-v-1-users-get-plain-v-10' // unused
```
And this is the output of bundle analyzer:
<img width="1267" alt="image" src="https://github.com/user-attachments/assets/b4a5920c-7bed-4f40-bce3-6f472cb05a82">

**Comment by remi-fongaufier** [comment]
I have the same issue, and I looking for documentation to find the best way to create a library, no buildable and no publishable, which is tree shakeable using esbuild or bun.

**Comment by lamhungypl** [comment]
Do you guys have any potential approaches?
